src/rust/src/sapling.rs

```diff
@@ -343,21 +366,51 @@ impl BatchValidator {
     /// `sighash` must be for the transaction this bundle is within.
     #[allow(clippy::boxed_local)]
     fn check_bundle(&mut self, bundle: Box<Bundle>, sighash: [u8; 32]) -> bool {
-        if let Some(validator) = &mut self.0 {
-            validator.check_bundle(bundle.0, sighash)
+        if let Some(inner) = &mut self.0 {
+            let cache = sapling_bundle_validity_cache();
+
+            // Compute the cache entry for this bundle.
+            let cache_entry = {
+                let bundle_commitment = bundle.commitment(TxIdDigester).unwrap();
+                let bundle_authorizing_commitment = bundle.commitment(BlockTxCommitmentDigester);
+                cache.compute_entry(
+                    bundle_commitment.as_bytes().try_into().unwrap(),
+                    bundle_authorizing_commitment.as_bytes().try_into().unwrap(),
+                    &sighash,
+                )
```

Comment:
This is the most likely place for a consensus bug, as I'm using the v5 bundle commitment logic for _all_ Sapling bundles. Reviewers should confirm that the Sapling cache entries do indeed bind the full bundle (we know that the sighash binds the entire v4 transaction's effects so that gets us part of the way if `bundle_commitment` were insufficient for v4, but it doesn't cover everything so `bundle_authorizing_commitment` is load-bearing here).

Reply:
I have verified that every part of the Sapling bundle that is bound in v4 sighashes is bound either in the txid or in the authorizing commitment. The one one difference between v4 sighashes and the auth bundle commitment is that spend_auth_sig is bound in the bundle commitment, whereas it was not previously bound (for obvious reasons) as part of the v4 sighash. I think that this actually means that we get commitments to the effects via both the bundle commitment and the authorizing commitment, which is fine.

Reply:
Because we're using the auth data commitment defined for v5, nothing binds Sprout authorizing data. But that's fine here. It would be an issue if we wanted to cache Sprout validation results.